File: src/logic/__tests__/BinaryConverter.test.ts
// @leosingleton/commonlibs - Common Libraries for TypeScript and .NET Core
// Copyright (c) Leo C. Singleton IV <leo@leosingleton.com>
// See LICENSE in the project root for license information.

import { BinaryConverter } from '../BinaryConverter';

describe('BinaryConverter', () => {

  it('UInt64', () => {
    const buffer = new Uint8Array(9);

    BinaryConverter.writeUInt64(buffer, 1, 0);
    const value1 = BinaryConverter.readUInt64(buffer, 1);
    expect(value1).toEqual(0);

    // JavaScript can't handle the largest unsigned 64-bit number, so go as high as it will
    BinaryConverter.writeUInt64(buffer, 1, Number.MAX_SAFE_INTEGER);
    const value2 = BinaryConverter.readUInt64(buffer, 1);
    expect(value2).toEqual(Number.MAX_SAFE_INTEGER);
  });

  it('Int32', () => {
    const buffer = new Uint8Array(5);

    BinaryConverter.writeInt32(buffer, 1, 0);
    const value1 = BinaryConverter.readInt32(buffer, 1);
    expect(value1).toEqual(0);

    BinaryConverter.writeInt32(buffer, 1, 2147483647);
    const value2 = BinaryConverter.readInt32(buffer, 1);
    expect(value2).toEqual(2147483647);

    BinaryConverter.writeInt32(buffer, 1, -2147483648);
    const value3 = BinaryConverter.readInt32(buffer, 1);
    expect(value3).toEqual(-2147483648);
  });

  it('UInt16', () => {
    const buffer = new Uint8Array(3);

    BinaryConverter.writeUInt16(buffer, 1, 0);
    const value1 = BinaryConverter.readUInt16(buffer, 1);
    expect(value1).toEqual(0);

    BinaryConverter.writeUInt16(buffer, 1, 0xffff);
    const value2 = BinaryConverter.readUInt16(buffer, 1);
    expect(value2).toEqual(0xffff);
  });

  it('Int16', () => {
    const buffer = new Uint8Array(3);

    BinaryConverter.writeInt16(buffer, 1, 0);
    const value1 = BinaryConverter.readInt16(buffer, 1);
    expect(value1).toEqual(0);

    BinaryConverter.writeInt16(buffer, 1, 32767);
    const value2 = BinaryConverter.readInt16(buffer, 1);
    expect(value2).toEqual(32767);

    BinaryConverter.writeInt16(buffer, 1, -32768);
    const value3 = BinaryConverter.readInt16(buffer, 1);
    expect(value3).toEqual(-32768);
  });

});

File: dotnet/CommonLibs/UnitTests/BinaryConverterTest.cs
using WhiteboardServer.Common;
using Xunit;

namespace WhiteboardServer.UnitTests.Common
{
    /// <summary>
    /// Tests the <see cref="BinaryConverter"/> class
    /// </summary>
    public class BinaryConverterTest
    {
        [Fact]
        public void UInt64()
        {
            var buffer = new byte[9];

            BinaryConverter.Write(buffer, 1, ulong.MinValue);
            var value1 = BinaryConverter.ReadUInt64(buffer, 1);
            Assert.Equal(ulong.MinValue, value1);

            BinaryConverter.Write(buffer, 1, ulong.MaxValue);
            var value2 = BinaryConverter.ReadUInt64(buffer, 1);
            Assert.Equal(ulong.MaxValue, value2);
        }

        [Fact]
        public void Int32()
        {
            var buffer = new byte[5];

            BinaryConverter.Write(buffer, 1, 0);
            var value1 = BinaryConverter.ReadInt32(buffer, 1);
            Assert.Equal(0, value1);

            BinaryConverter.Write(buffer, 1, int.MinValue);
            var value2 = BinaryConverter.ReadInt32(buffer, 1);
            Assert.Equal(int.MinValue, value2);

            BinaryConverter.Write(buffer, 1, int.MaxValue);
            var value3 = BinaryConverter.ReadInt32(buffer, 1);
            Assert.Equal(int.MaxValue, value3);
        }

        [Fact]
        public void UInt16()
        {
            var buffer = new byte[3];

            BinaryConverter.Write(buffer, 1, ushort.MinValue);
            var value1 = BinaryConverter.ReadUInt16(buffer, 1);
            Assert.Equal(ushort.MinValue, value1);

            BinaryConverter.Write(buffer, 1, ushort.MaxValue);
            var value2 = BinaryConverter.ReadUInt16(buffer, 1);
            Assert.Equal(ushort.MaxValue, value2);
        }

        [Fact]
        public void Int16()
        {
            var buffer = new byte[3];

            BinaryConverter.Write(buffer, 1, (short)0);
            var value1 = BinaryConverter.ReadInt16(buffer, 1);
            Assert.Equal(0, value1);

            BinaryConverter.Write(buffer, 1, short.MinValue);
            var value2 = BinaryConverter.ReadInt16(buffer, 1);
            Assert.Equal(short.MinValue, value2);

            BinaryConverter.Write(buffer, 1, short.MaxValue);
            var value3 = BinaryConverter.ReadInt16(buffer, 1);
            Assert.Equal(short.MaxValue, value3);
        }
    }
}
